superbench/analyzer/data_diagnosis.py

```diff
@@ -0,0 +1,359 @@
+                return True
+        return False
+
+    def single_node_diagnosis(self, node):
+        """Use rules to diagnosis single node data.
+
+        Use the rules defined in rule_file to diagnose the raw data of each node,
+        if the node violate any rule, label as issued node and save
+        the 'Hw Issues', '# of Issues', 'Category', 'Issue Details' and processed data of issued node.
+
+        Args:
+            node (str): the node to do the diagosis
+
+        Returns:
+            None if the node is not labeled as issued,
+            otherwise return details for the issued node.
+        """
+        data_row = self._raw_data_df.loc[node]
+        columns = self._raw_data_df.columns
+        issue_label = False
+        category_details = []
+        categories = set()
+        summary_data_row = pd.Series(index=self._sb_baseline.keys(), name=node, dtype=float)
+        cnn_benchmarks = ['densenet_models', 'vgg_models', 'resnet_models']
+        model_label_num = 0
+
+        for metric in self._sb_baseline:
+            benchmark = metric.split('/')[0]
+            baseline = self._sb_baseline[metric]['criteria']
+            pass_rule = True
+            # metric not in raw_data not the value is none, miss test
+            if metric not in columns or pd.isna(data_row[metric]):
+                pass_rule = False
+            # check if metric pass the rule
+            else:
+                data = data_row[metric]
+                rule = self._sb_baseline[metric]['rules']
+                rule_op = RuleOp.get_rule_func(RuleType(rule['name']))
+                (pass_rule, processed) = rule_op(data, baseline, rule)
+                summary_data_row[metric] = processed
+            # label the node as issued one
+            if not pass_rule:
+                # use return code to identify 'miss test'
+                if 'return_code' in metric:
+                    category_details.append(benchmark + '_miss')
+                    categories.add('MissTest')
+                    issue_label = True
+                else:
+                    category_details.append(metric)
+                    categories.add(benchmark)
+                    if benchmark not in cnn_benchmarks:
+                        issue_label = True
+                    # for cnn models, only 2 model metric issued, label as issue
+                    else:
+                        model_label_num += 1
+                        if model_label_num >= 2:
+                            issue_label = True
+
+        if issue_label:
+            # Add category information
+            general_cat_str = ','.join(categories)
+            details_cat_str = ','.join(category_details)
+            hw_issue_flag = self.hw_issue(categories)
+            details_row = [hw_issue_flag, len(category_details), general_cat_str, details_cat_str]
+            return (details_row, summary_data_row)
+
+        return (None, None)
+
+    def rule_based_diagnosis(self, rule_file):
```

Comment:
Do you have other based diagnosis? Why we want to highlight rule-based? Why not `run_diagnosis_rules()`?

Reply:
updated